<a href="https://colab.research.google.com/github/anish-git-05/Linear_Regression_House_Prediction/blob/main/House_Rent_Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd
import numpy as np

In [3]:
df=pd.read_csv('House_Rent_Dataset.csv')
df.head()

,Posted On,BHK,Rent,Size,Floor,Area Type,Area Locality,City,Furnishing Status,Tenant Preferred,Bathroom,Point of Contact
0,2022-05-18,2,10000,1100,Ground out of 2,Super Area,Bandel,Kolkata,Unfurnished,Bachelors/Family,2,Contact Owner
1,2022-05-13,2,20000,800,1 out of 3,Super Area,"Phool Bagan, Kankurgachi",Kolkata,Semi-Furnished,Bachelors/Family,1,Contact Owner
2,2022-05-16,2,17000,1000,1 out of 3,Super Area,Salt Lake City Sector 2,Kolkata,Semi-Furnished,Bachelors/Family,1,Contact Owner
3,2022-07-04,2,10000,800,1 out of 2,Super Area,Dumdum Park,Kolkata,Unfurnished,Bachelors/Family,1,Contact Owner
4,2022-05-09,2,7500,850,1 out of 2,Carpet Area,South Dum Dum,Kolkata,Unfurnished,Bachelors,1,Contact Owner


In [4]:
df.tail()

,Posted On,BHK,Rent,Size,Floor,Area Type,Area Locality,City,Furnishing Status,Tenant Preferred,Bathroom,Point of Contact
4741,2022-05-18,2,15000,1000,3 out of 5,Carpet Area,Bandam Kommu,Hyderabad,Semi-Furnished,Bachelors/Family,2,Contact Owner
4742,2022-05-15,3,29000,2000,1 out of 4,Super Area,"Manikonda, Hyderabad",Hyderabad,Semi-Furnished,Bachelors/Family,3,Contact Owner
4743,2022-07-10,3,35000,1750,3 out of 5,Carpet Area,"Himayath Nagar, NH 7",Hyderabad,Semi-Furnished,Bachelors/Family,3,Contact Agent
4744,2022-07-06,3,45000,1500,23 out of 34,Carpet Area,Gachibowli,Hyderabad,Semi-Furnished,Family,2,Contact Agent
4745,2022-05-04,2,15000,1000,4 out of 5,Carpet Area,Suchitra Circle,Hyderabad,Unfurnished,Bachelors,2,Contact Owner


In [5]:
furnishing_map={"Unfurnished":1,"Semi-Furnished":2,"Furnished":3}
df['Furnishing_val']=df["Furnishing Status"].map(furnishing_map)

We have created an extra column which quantifies the furnishing status.

In [6]:
df.head()

,Posted On,BHK,Rent,Size,Floor,Area Type,Area Locality,City,Furnishing Status,Tenant Preferred,Bathroom,Point of Contact,Furnishing_val
0,2022-05-18,2,10000,1100,Ground out of 2,Super Area,Bandel,Kolkata,Unfurnished,Bachelors/Family,2,Contact Owner,1
1,2022-05-13,2,20000,800,1 out of 3,Super Area,"Phool Bagan, Kankurgachi",Kolkata,Semi-Furnished,Bachelors/Family,1,Contact Owner,2
2,2022-05-16,2,17000,1000,1 out of 3,Super Area,Salt Lake City Sector 2,Kolkata,Semi-Furnished,Bachelors/Family,1,Contact Owner,2
3,2022-07-04,2,10000,800,1 out of 2,Super Area,Dumdum Park,Kolkata,Unfurnished,Bachelors/Family,1,Contact Owner,1
4,2022-05-09,2,7500,850,1 out of 2,Carpet Area,South Dum Dum,Kolkata,Unfurnished,Bachelors,1,Contact Owner,1


In [7]:
X=df[['BHK','Size','Bathroom','Furnishing_val']].to_numpy()

In [8]:
y=df['Rent'].to_numpy()
y

array([10000, 20000, 17000, ..., 35000, 45000, 15000], shape=(4746,))

In [9]:
col_mean=np.mean(X,axis=0)
col_sd=np.std(X,axis=0)

In [10]:
np.save('mean.npy',col_mean)
np.save('std.npy',col_sd)

In [11]:
X_scaled=(X-col_mean)/col_sd

In [12]:
print(np.mean(X_scaled,axis=0))
print(np.std(X_scaled,axis=0))

[ 2.99428039e-17  8.98284116e-17 -1.19771215e-16 -5.98856077e-17]
[1. 1. 1. 1.]


The mean for all columns is close to 0 and the standard deviation is close to 1.

In [13]:
w=np.zeros(X_scaled.shape[1])
b=0.0
initial_loss=np.mean((X_scaled@w+b-y)**2)
def gradient_descent(X,y,w,b,alpha=0.01,iters=1000):
  n=X.shape[0]
  for i in range(iters):
    y_cap=X@w+b
    error=y_cap-y
    dw=(2/n)*(X.T@error)
    db=(2/n)*(np.sum(error))
    w=w-alpha*dw
    b=b-alpha*db
  return w,b

w_new,b_new=gradient_descent(X_scaled,y,w,b,alpha=0.01,iters=1000)
print(w_new)
print(b_new)

[ -884.57455016 14930.95420674 23195.57856651  4760.01951396]
34993.45126854078


In [14]:
final_loss=np.mean((X_scaled@w_new+b_new)**2)
print(np.sqrt(initial_loss))
print(np.sqrt(final_loss))

85579.60011574313
50337.99554321507


In [15]:
print(X_scaled@w_new+b_new)

[33806.66545279  7470.92753785 12180.00122352 ... 81228.71932608
 49116.04350224 31452.12860996]


In [16]:
cities=list(df['City'].unique())
cities

['Kolkata', 'Mumbai', 'Bangalore', 'Delhi', 'Chennai', 'Hyderabad']

We have to take into account the city of rent. So we have done the binary encoding of each city present in the data.

In [17]:
df['Is_Kolkata']=(df['City']=='Kolkata').astype(int)
df['Is_Delhi']=(df['City']=='Delhi').astype(int)
df['Is_Mumbai']=(df['City']=='Mumbai').astype(int)
df['Is_Chennai']=(df['City']=='Chennai').astype(int)
df['Is_Bangalore']=(df['City']=='Bangalore').astype(int)
df['Is_Hyderabad']=(df['City']=='Hyderabad').astype(int)
X_num=df[['BHK','Size','Bathroom','Furnishing_val']]
X_city=df[['Is_Kolkata','Is_Delhi','Is_Mumbai','Is_Chennai','Is_Bangalore','Is_Hyderabad']].to_numpy()

In [18]:
col_mean=np.mean(X_num,axis=0)
col_sd=np.std(X_num,axis=0)
X_scaled=(X_num-col_mean)/(col_sd)
X_final=np.concatenate((X_scaled,X_city),axis=1)

Only scaled the numeical continuous values. Didn't scale the discrete binary indicators.

In [19]:
w=np.zeros(X_final.shape[1])
b=0.0
w_new,b_new=gradient_descent(X_final,y,w,b,alpha=0.001)
new_loss=np.mean((X_final@w_new+b_new-y)**2)
print(np.sqrt(new_loss))
print(np.mean(y))
y_pred=X_final@w_new+b_new

67142.03545846639
34993.45132743363


In [20]:
np.save('w.npy',w_new)
np.save('b.npy',b_new)

In [21]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
lr=LinearRegression()
lr.fit(X_final,y)
y_pred=lr.predict(X_final)
print(np.sqrt(mean_squared_error(y,y_pred)))
print(np.mean(y))

64735.125121586774
34993.45132743363


We can see there is not much difference between the Root mean square error of gradient descent approach and by using the built in function of sklearn library.

From gradient descent: 67142.03545846639
From sklearn library: 64735.125121586774